Added support for secrets and databases to serverless datacenter

davidthor opened the pull request.

davidthor commented:
@TylerAldrich I'm trying to get a deployment of my twitter clone app working and am getting a weird error. Have you seen this before?

```
% arcctl deploy architectio/twitter-clone:latest -e dthor -i app:app
Error: There is already a Construct with name 'dthor-architectio-kratos-kratos-deployment' in GoogleCloudDeploymentModule [arc-architectio--kratos--deployment--kratos-blue]
    at o.addChild (https://esm.sh/v124/constructs@10.2.27/denonext/constructs.mjs:4:3339)
    at new o (https://esm.sh/v124/constructs@10.2.27/denonext/constructs.mjs:4:574)
    at new f (https://esm.sh/v124/constructs@10.2.27/denonext/constructs.mjs:4:3906)
    at new dr (https://esm.sh/v124/cdktf@0.15.5/denonext/cdktf.mjs:15:5447)
    at new Mr (https://esm.sh/v124/cdktf@0.15.5/denonext/cdktf.mjs:18:1969)
    at new CloudRunV2Service (file:///Users/davidthor/Workspaces/arc-v2/arcctl/src/@providers/gcp/.gen/providers/google/cloud-run-v2-service/index.ts:4314:5)
    at new GoogleCloudDeploymentModule (file:///Users/davidthor/Workspaces/arc-v2/arcctl/src/@providers/gcp/modules/deployment.ts:46:26)
    at CldCtlTerraformStack.addModule (file:///Users/davidthor/Workspaces/arc-v2/arcctl/src/utils/stack.ts:15:20)
    at GoogleCloudDeploymentService.getHash (file:///Users/davidthor/Workspaces/arc-v2/arcctl/src/@providers/terraform.service.ts:510:11)
    at PipelineStep.getHash (file:///Users/davidthor/Workspaces/arc-v2/arcctl/src/pipeline/step.ts:93:28)
```

TylerAldrich commented:
> @TylerAldrich I'm trying to get a deployment of my twitter clone app working and am getting a weird error. Have you seen this before?
> 
> ```
> % arcctl deploy architectio/twitter-clone:latest -e dthor -i app:app
> Error: There is already a Construct with name 'dthor-architectio-kratos-kratos-deployment' in GoogleCloudDeploymentModule [arc-architectio--kratos--deployment--kratos-blue]
>     at o.addChild (https://esm.sh/v124/constructs@10.2.27/denonext/constructs.mjs:4:3339)
>     at new o (https://esm.sh/v124/constructs@10.2.27/denonext/constructs.mjs:4:574)
>     at new f (https://esm.sh/v124/constructs@10.2.27/denonext/constructs.mjs:4:3906)
>     at new dr (https://esm.sh/v124/cdktf@0.15.5/denonext/cdktf.mjs:15:5447)
>     at new Mr (https://esm.sh/v124/cdktf@0.15.5/denonext/cdktf.mjs:18:1969)
>     at new CloudRunV2Service (file:///Users/davidthor/Workspaces/arc-v2/arcctl/src/@providers/gcp/.gen/providers/google/cloud-run-v2-service/index.ts:4314:5)
>     at new GoogleCloudDeploymentModule (file:///Users/davidthor/Workspaces/arc-v2/arcctl/src/@providers/gcp/modules/deployment.ts:46:26)
>     at CldCtlTerraformStack.addModule (file:///Users/davidthor/Workspaces/arc-v2/arcctl/src/utils/stack.ts:15:20)
>     at GoogleCloudDeploymentService.getHash (file:///Users/davidthor/Workspaces/arc-v2/arcctl/src/@providers/terraform.service.ts:510:11)
>     at PipelineStep.getHash (file:///Users/davidthor/Workspaces/arc-v2/arcctl/src/pipeline/step.ts:93:28)
> ```

I haven't seen that but it looks like that means the name I was giving the `Deployment` is probably not as unique as I thought it would be... and I actually just thought about why. I removed the port number because I thought it was superfluous after making other changes, but just now realized I added it in the first place in case you deployed the same service with multiple ports exposed. That's likely why you're seeing this so I'll need to add that back 🤦

davidthor commented:
> I haven't seen that but it looks like that means the name I was giving the Deployment is probably not as unique as I thought it would be... and I actually just thought about why. I removed the port number because I thought it was superfluous after making other changes, but just now realized I added it in the first place in case you deployed the same service with multiple ports exposed. That's likely why you're seeing this so I'll need to add that back 🤦

Makes sense! I included some fixes for those issues in this PR. This is going strait into your branch so let me know if there's any chance you want.

TylerAldrich commented on src/@providers/gcp/modules/deployment.ts:
```diff
@@ -40,12 +40,13 @@ export class GoogleCloudDeploymentModule extends ResourceModule<
-        (this.inputs?.name.replaceAll('/', '-') || 'deleting') + `-${service.port || 80}`;
-      const deployment = new CloudRunV2Service(this, `${deployment_name}-deployment`, {
+        (this.inputs?.name.replaceAll('/', '-') || 'deleting');
+      const resource_name = `${deployment_name}-svc-${index}`;
```
I used `service.port` before because we need to also construct this name in the `service` module where we don't actually know which index it would be in the list - so if we update this to 

```
const resource_name = `${deployment_name}-svc-${service.port || 80}`;
```

we could change https://github.com/architect-team/arcctl/pull/89/files#diff-d16a7edc1c8f99048a9c7e26b6786900571e2e7165997e82c80bf1eb2afdfd64R22
to match

TylerAldrich commented on src/pipeline/pipeline.ts:
```diff
@@ -256,6 +256,7 @@ export class Pipeline {
       if (!this.steps.some((n) => n.id === edge.from)) {
         throw new Error(`${edge.from} is missing from the pipeline`);
       } else if (!this.steps.some((n) => n.id === edge.to)) {
+        console.log(this.steps.filter((s) => s.type === 'databaseCluster'));
```
probably can 🔪 this too

TylerAldrich commented:
> > I haven't seen that but it looks like that means the name I was giving the Deployment is probably not as unique as I thought it would be... and I actually just thought about why. I removed the port number because I thought it was superfluous after making other changes, but just now realized I added it in the first place in case you deployed the same service with multiple ports exposed. That's likely why you're seeing this so I'll need to add that back 🤦
> 
> Makes sense! I included some fixes for those issues in this PR. This is going strait into your branch so let me know if there's any chance you want.

Everything here looks good to me except the 1 thing about port vs index number in the CloudRun function name, otherwise this is good to go

davidthor commented:
> Everything here looks good to me except the 1 thing about port vs index number in the CloudRun function name, otherwise this is good to go

Updated! Merge it into your branch when you're ready.
